File: fish-war/src/cli/playGame.ts
#!/usr/bin/env node
import { BoardGameLogic } from '../utils/boardGameLogic';
import { createInterface } from 'readline';

const rl = createInterface({
  input: process.stdin,
  output: process.stdout
});

const game = new BoardGameLogic();

function displayGame() {
  console.clear();
  console.log('=================================');
  console.log('       おさかな対戦 CLI');
  console.log('=================================\n');
  console.log(game.getBoardText());
}

function askForMove(): Promise<string> {
  return new Promise((resolve) => {
    rl.question('手を入力してください (例: い↑B3B2, 終了: quit): ', (answer) => {
      resolve(answer.trim());
    });
  });
}

async function playGame() {
  displayGame();

  while (!game.getGameState().gameResult) {
    const moveString = await askForMove();
    
    if (moveString.toLowerCase() === 'quit') {
      console.log('\nゲームを終了します。');
      rl.close();
      return;
    }

    const move = game.parseMove(moveString);
    if (!move) {
      console.log('\n❌ エラー: 無効な手の形式です。例: い↑B3B2');
      console.log('続けるにはEnterキーを押してください...');
      await askForMove();
      displayGame();
      continue;
    }

    const result = game.makeMove(move);
    if (!result.success) {
      console.log(`\n❌ エラー: ${result.error}`);
      console.log('続けるにはEnterキーを押してください...');
      await askForMove();
      displayGame();
      continue;
    }

    displayGame();

    if (result.gameResult) {
      console.log('\n=================================');
      if (result.gameResult.type === 'win') {
        const winner = result.gameResult.winner === 'first' ? '先手' : '後手';
        console.log(`🎉🎉🎉 ${winner}の勝利！ 🎉🎉🎉`);
        console.log(`理由: ${result.gameResult.reason}`);
      } else if (result.gameResult.type === 'draw') {
        console.log(`🤝 引き分け`);
        console.log(`理由: ${result.gameResult.reason}`);
      }
      console.log('=================================\n');
      
      const playAgain = await new Promise<string>((resolve) => {
        rl.question('もう一度プレイしますか？ (y/n): ', resolve);
      });
      
      if (playAgain.toLowerCase() === 'y') {
        // 新しいゲームを開始
        const newGame = new BoardGameLogic();
        Object.assign(game, newGame);
        displayGame();
      } else {
        console.log('ゲームを終了します。');
        rl.close();
        return;
      }
    }
  }

  rl.close();
}

// ゲーム開始
console.log('おさかな対戦へようこそ！\n');
console.log('ルール:');
console.log('- 相手のまぐろを捕獲すると勝利');
console.log('- 自分のまぐろが相手陣地に到達すると勝利');
console.log('- 手の入力例: い↑B3B2 (コマを移動), い↑A3★ (手ゴマを配置)');
console.log('- 終了するには "quit" と入力\n');
console.log('Enterキーを押してゲームを開始...');

rl.question('', () => {
  playGame().catch((err) => {
    console.error('エラーが発生しました:', err);
    rl.close();
  });
});

File: fish-war/python/src/play_game.py
#!/usr/bin/env python3
"""
おさかな対戦 - インタラクティブCLI版
"""

import sys
from board_game_logic import (
    initialize_game, get_board_text, get_game_info_text,
    parse_move, make_move
)


def print_game_state(game_state):
    """ゲーム状態を表示"""
    print("\n" + "=" * 30)
    print(get_board_text(game_state.board))
    print("-" * 30)
    print(get_game_info_text(game_state))
    print("=" * 30)


def print_help():
    """ヘルプメッセージを表示"""
    print("\n=== 操作方法 ===")
    print("移動: コマ種類+方向+移動元+移動先 (例: い↑B3B2)")
    print("配置: コマ種類+*+配置先 (例: い*B2)")
    print("コマ: ま=まぐろ, い=いなだ, ぶ=ぶり, た=たこ, か=かれい")
    print("方向: ↑↓→← (実際の移動方向とは関係なく入力)")
    print("help: このヘルプを表示")
    print("quit: ゲームを終了")
    print("================\n")


def main():
    print("おさかな対戦へようこそ！")
    print("'help'でヘルプを表示します。")
    
    game_state = initialize_game()
    print_game_state(game_state)
    
    while game_state.game_result is None:
        player = "先手" if game_state.current_player == 'first' else "後手"
        
        try:
            move_input = input(f"\n{player}の番です。手を入力してください: ").strip()
            
            if move_input.lower() == 'quit':
                print("ゲームを終了します。")
                sys.exit(0)
            elif move_input.lower() == 'help':
                print_help()
                continue
            
            # 手を解析
            move = parse_move(move_input)
            if not move:
                print("エラー: 入力形式が正しくありません。'help'でヘルプを表示できます。")
                continue
            
            # 移動を実行
            new_state, error = make_move(game_state, move)
            if error:
                print(f"エラー: {error}")
                continue
            
            game_state = new_state
            print_game_state(game_state)
            
        except KeyboardInterrupt:
            print("\n\nゲームを中断しました。")
            sys.exit(0)
        except Exception as e:
            print(f"予期しないエラーが発生しました: {e}")
            continue
    
    # ゲーム終了
    print("\n" + "★" * 30)
    if game_state.game_result == 'draw':
        print("引き分けです！")
    else:
        winner = "先手" if game_state.game_result == 'first' else "後手"
        print(f"{winner}の勝利です！")
    print("★" * 30)


if __name__ == "__main__":
    main()